tools/online_editor/src/proxy.ts

```diff
@@ -0,0 +1,65 @@
+// Copyright © SixtyFPS GmbH <info@slint-ui.com>
+// SPDX-License-Identifier: GPL-3.0-only OR LicenseRef-Slint-commercial
+
+import {
+  MessageReader,
+  MessageWriter,
+  DataCallback,
+  Message,
+  Disposable,
+  Event,
+  PartialMessageInfo,
+} from "monaco-languageclient";
+
+export class ProxyReader implements MessageReader {
+  constructor(reader: MessageReader, callback: (_: Message) => boolean) {
+    this.#reader = reader;
+    this.#callback = callback;
+    this.onError = this.#reader.onError;
+    this.onClose = this.#reader.onClose;
+    this.onPartialMessage = this.#reader.onPartialMessage;
+  }
+
+  readonly #reader: MessageReader;
+  readonly #callback: (_: Message) => boolean;
+  onError: Event<Error>;
+  onClose: Event<void>;
+  onPartialMessage: Event<PartialMessageInfo>;
+
+  dispose() {
+    this.#reader.dispose();
+  }
+
+  listen(callback: DataCallback): Disposable {
+    return this.#reader.listen((data: Message) => {
+      this.#callback(data) || callback(data);
+    });
+  }
+}
+
+export class ProxyWriter implements MessageWriter {
+  constructor(writer: MessageWriter, callback: (_: Message) => void) {
+    this.#writer = writer;
+    this.#callback = callback;
+    this.onError = this.#writer.onError;
+    this.onClose = this.#writer.onClose;
+  }
+
+  readonly #writer: MessageWriter;
+  readonly #callback: (_: Message) => void;
+  onError: Event<[Error, Message | undefined, number | undefined]>;
+  onClose: Event<void>;
+
+  write(msg: Message): Promise<void> {
+    this.#callback(msg);
+    return this.#writer.write(msg);
+  }
+
+  end(): void {
+    this.#writer.end();
+  }
+
+  dispose(): void {
+    this.#writer.dispose();
+  }
+}
```

Comment:
It looks like this class is neither used nor needed.

```suggestion
```

Reply:
Yes, I thought I need to inject the reply somehow, but turns out that was not needed.

I thought it does not hurt to keep it around:-) But you are right: The code is too simple to keep just in case.

Reply:
Gone now...